plugin.video.massengeschmack/resources/lib/listing.py

```diff
@@ -0,0 +1,146 @@
+# -*- coding: utf-8 -*-
+#
+# Massengeschmack Kodi add-on
+# Copyright (C) 2013-2016 by Janek Bevendorff
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import xbmcplugin
+import xbmcgui
+
+from globalvars import *
+
+
+class Listing(object):
+    def __init__(self):
+        self.__source = None
+
+    def generate(self, source):
+        """
+        Generate listing from data source.
+        Will return False if the given DataSource does not intend to return a non-empty listing generator,
+        i.e. when L{resources.lib.datasource.DataSource.hasListItems} returns False
+
+        @type source: resources.lib.datasource.DataSource
+        @param source: the data source object
+        @rtype: bool
+        @return Whether the given DataSource actually intends to return a non-empty listing
+        """
+        if not source.hasListItems():
+            return False
+
+        self.__source = source
+        items         = source.getListItems()
+
+        for i in items:
+            self.__addDir(i)
+
+        return True
+
+    def show(self):
+        """
+        Show the listing after it has been generated.
+        You should check the output of L{generate} before trying to show a listing.
+        """
+        if 'true' == ADDON.getSetting('advanced.viewmodeFix'):
+            self.setViewMode(510)
+        else:
+            self.setViewMode(self.__source.getViewMode())
+        xbmcplugin.endOfDirectory(ADDON_HANDLE)
+
+    def setViewMode(self, id):
+        """
+        Set the view mode of the current listing.
+
+        @type id: int
+        @param id: the view mode ID from the current skin
+        """
+        if 'false' == ADDON.getSetting('advanced.adjustViewModes'):
+            return
+
+        xbmcplugin.setContent(ADDON_HANDLE, self.__source.getContentMode())
+        xbmc.executebuiltin('Container.SetViewMode(' + str(id) + ')')
```

Comment:
This isn't allowed.

Reply:
I've discussed this countless times with reviewers. Some of your rules enforce extremely bad usability (like this one, were not allowing for sensible view modes just drops the user back into the unusable default every time they navigate one page further in the listing). This is already configurable just so as to meet your requirements, although I'd expect 90%+ of the users to leave it on.

Reply:
Perhaps to add to that, this would not be necessary if Kodi

- allowed to set scopes of pages that share the same view mode setting
- allow plugins to define a default view mode for each scope which may then be changed by the user

Right now I can only enforce a view mode or not enforce it, which would a) give the user a terrible experience at first launch and b) give them an even more terrible experience when they paginate, because every new page resets itself to the default viewmode.